Sources/Raft/Consensus/Consensus.swift

```diff
@@ -0,0 +1,227 @@
+        if _tryMoveTo(nextState: .preCandidate) {
+            return .startPreVote
+        }
+        return .scheduleNextTimer(delay: electionTimeoutInterval())
+    }
+
+    public enum VoteType {
+        case preVote
+        case vote
+    }
+
+    /// Try to run a pre vote campaign
+    public func startPreVote() async -> ElectionCommand {
+        let term = self.term
+        let voteResult = await startVote(type: .preVote)
+        if voteResult {
+            logger.debug("Won preVote election, starting vote")
+            return .startVote
+        }
+        logger.warning("Lost preVote for \(term)")
+        return .scheduleNextTimer(delay: electionTimeoutInterval())
+    }
+
+    /// Try to run a election campaign
+    public func startVote() async -> ElectionCommand {
+        let voteResult = await startVote(type: .preVote)
+        logger.debug("Finish campign", metadata: [
+            "vote/result": "\(voteResult)",
+            "vote/term": "\(self.term)"
+        ])
+        if voteResult && _tryMoveTo(nextState: .leader) {
+            return .startToBeALeader
+        }
+        self.logger.debug("Failed to become a leader for \(self.term) term")
+        return .scheduleNextTimer(delay: electionTimeoutInterval())
+    }
+
+    /// Node is ready to start new vote
+    func startVote(type: VoteType) async -> Bool {
+        let term = self.term.nextTerm()
+        if case .vote = type {
+            // Set new term if this is a real vote
+            self.term = term
+        }
+        // How much votes we need to win
+        let tallyVotes = self.peers.quorumSize
+        let request = RequestVote.Request(type: type,
+                                          term: term.id,
+                                          candidate: myself,
+                                          lastLogIndex: 0,
+                                          lastLogTerm: 0)
+        // Get list of active peers
+        let peers = self.peers
+
+        // TODO Handle errors correctly
+        let result = try! await Task.withGroup(resultType: Bool.self, returning: Bool.self) { group in
+
+            // We should stop election at the moment when we got quorum
+            for peer in peers {
+                // We should stop election at the moment when we got quorum
+                await group.add(operation: {
+                    await peer.requestVote(request).voteGranted
+                })
+            }
+
+            do {
+                var grantedVotes: UInt = 1 // We already votes for ourself
+                while let result = try await group.next() {
+                    grantedVotes += result ? 1 : 0
+                    if grantedVotes >= tallyVotes {
+                        return true
+                    }
+                }
+                return false
+            } catch {
+                // If got any error we lost an election
+                return false
```

Comment:
minor thing but you could `group.cancelAll()` before returning here... It won't really be able to do much if the requests are already shot off, but if they're not the tasks are cancelled and we can avoid unnecessary work

Reply:
Will also add cancellation to the case when we already got enough voces